doc/bakugo_spec.md

```diff
@@ -0,0 +1,1111 @@
+
+## Type identity
+Two types are either _identical_ or _different_.
+
+A _named type_ is always different from any other type. Otherwise, two types are identical if their
+_underlying_ type literals are structurally equivalent; that is, they have the same literal structure and
+corresponding components have identical types.
+```
+type (
+	A0 = float32
+	A1 = A0
+	A2 = bool
+	A3 = int
+
+	B0 A0
+	B1 float32
+
+	C0 = B0
+)
+```
+these types are identical:
+```
+A0, A1, and float32
+A2 and bool
+A3 and int
+
+B0 and C0
+int and int
+```
+
+B0 and B1 are different because they are new types created by distinct type definitions.
+
+## Assignability
+A value x of type V is _assignable_ to a _variable_ of type T ("x is assignable to T") if one of the
+following conditions applies:
+
+- V and T are identical.
+- V and T have identical _underlying types_ and atleast one of V or T is not a _named type_.
+- x is an untyped _constant representable_ by a value of type T.
+
+## Representability
+A _constant_ x is _representable_ by a value of tye T, if one of the following conditions applies:
+
+- x is in the set of value _determined_ by T 
+- T is a floating-point type and x can be rounded to T's precision without overflow. Rounding uses IEEE
+754 round-to-even rules but with an IEEE negative zero further simplified to an unsigned zero. Note that
+constant values never result in an IEEE negative zero, NaN, or infinity.
+
+```
+x                   T           x is representable by a value of T because
+
+1024                int16       1024 is in the set of 16-bit integers
+1e10                uint64      10000000000 is in the set of unsigned 64-bit integers
+2.718281828459045   float32     2.718281828459045 rounds to 2.7182817 which is in the set of float32 values
+-1e-1000            float64     -1e-1000 rounds to IEEE -0.0 which is further simplified to 0.0
+```
+```
+x                   T           x is not representable by a value of T because
+
+0                   bool        0 is not in the set of boolean values
+-1                  uint16      -1 is not in the set of unsigned 16-bit integers
+```
+
+# Blocks
+
+A block is a possibly empty sequence of declarations and statements within matching brace brackets.
+```
+Block = "{" StatementList "}" .
+StatementList = { Statement ";" } .
+```
+
+In addition to explicit blocks in the source code, there are implicit blocks:
+
+1. The _universe block_ encompasses all Go source text.
+2. Each _package_ has a _package block_ containing all Go source text for that package.
+3. Each file has a _file block_ containing all Go source text in that file.
+4. _"if"_ statement is considered to be in it's own implicit block
+
+Blocks nest and influence _scoping_.
+
+# Declarations and scope
+
+A _declaration_ binds a non-_blank_ identifier to a _constant_, _type_, _variable_, _function_. Every
+identifier in a program must be declared. No identifier may be declared twice in the same block, and no
```

Comment:
An extension worth considering here: allow re-declaring an identifier of the same name again in a block.

This is possible in Rust and is actually useful.

See this example pseudocode:
```
count := "1234"  // is a string here
count := parse(count)  // is a uint here
count := SomeCustomCountStruct { count: count }  // is a custom type here

// this also ensures that the string count and uint count aren't used beyond this point
```

Vs the way it is done in Go currently:
```
countStr := "1234"
countParsed := parse(count)
count := SomeCustomCountStruct { count: count }

// countStr and countParsed can be used here even though that may be undesired
```

---

Implementing this might be difficult though. Scopes will have to be non-lexical!

Reply:
I agree with this proposal and think it would be a valuable feature to have. I can relate to it from OCaml. Implementing might be challenging, but I am eager to see it become a reality.

Good one!